specs/plugins.md

```diff
@@ -0,0 +1,134 @@
+
+## Plugins
+
+Plugin classes must implement the `IProxyPlugin` interface and provide a parameterless constructor.
+
+```cs
+public interface IProxyPlugin {
+    string Name { get; }
+    void Register(IPluginEvents pluginEvents, IConfigurationSection? configSection);
+}
+
+public interface IPluginEvents {
+    event EventHandler<InitArgs> Init;
+    event EventHandler<OptionsLoadedArgs> OptionsLoaded;
+    event EventHandler<ProxyRequestArgs> Request;
+    event EventHandler<ProxyResponseArgs> Response;
+}
+```
+
+### Registration
+
+The `Register` method of a `IProxyPlugin` will be called after an instance is created and will provide an `ISet<Regex>` to define the urls the plugin wll watch, either using the default or a plugin specific set based on the supplied configuration and an `IConfigurationSection` if one was loaded.
+
+> It is strongly recommended that plugin implementers provide a default configuration in their code.
+
+Plugin classes may register a handler for any events provided in the `IPluginEvents` interface. Plugins registering a handler for the `Init` event should also register a handler for the `OptionsLoaded` event to consume the supplied options.
+
+All of the event arg types will inherit from a common base type which will supply common logic and utilities. A proposed implementation of the `PluginArgsBase` is provided here:
+
+```cs
+public abstract class PluginArgsBase {
+    protected PluginArgsBase(IConsoleLogger consoleLogger) {
+        ConsoleLogger = consoleLogger ?? throw new ArgumentNullException(nameof(consoleLogger));
+    }
+    public IConsoleLogger ConsoleLogger { get; set; }
+
+    public static bool IsSdkRequest(Request request) {
+        return request.Headers.HeaderExists("SdkVersion");
+    }
+
+    public static bool IsGraphRequest(Request request) {
+        return request.RequestUri.Host.Contains("graph", StringComparison.OrdinalIgnoreCase);
+    }
+    public static string GetMoveToSdkUrl(Request request) {
+        // TODO: return language-specific guidance links based on the language detected from the User-Agent
+        return "https://aka.ms/move-to-graph-js-sdk";
+    }
+
+    public static string BuildApiErrorMessage(Request r) => $"Some error was generated by the proxy. {(IsGraphRequest(r) ? (IsSdkRequest(r) ? "" : BuildUseSdkMessage(r)) : "")}";
+
+    public static string BuildUseSdkMessage(Request r) => $"To handle API errors more easily, use the Graph SDK. More info at {GetMoveToSdkUrl(r)}";
+}
+```
+
+The details of the IConsoleLogger interface and implementation are out of the scope of this spec, except to say that they will provide plugins with a standardized mechanism for logging messages using preset formatting and categorization.
```

Comment:
Good point! For reference: we'd implement something like https://github.com/waldekmastykarz/console-output/blob/135c9751ffedbe1737d20f7f55910b0d50fa7439/Program.cs#L51 and then call it passing the message to log and its type. In the method, we should also allow passing multiple strings to allow printing multi-line messages.